[DataLoader] Make distributed lazily initialized & share seed via PG

ejguan opened the pull request:
Fixes #84492 https://github.com/pytorch/data/issues/772

## Changes
- Move the logic of distributed sharding from the constructor of DataLoader to the constructor of DataLoaderIterator. This would prevent the Error caused by lazy distributed process initialization
- Replace distributed store by process group (`gloo`) to share the random seed because `mpi` backend doesn't provide distributed store.

ejguan commented on torch/utils/data/dataloader.py:
```diff
@@ -631,6 +575,9 @@ def __init__(self, loader: DataLoader) -> None:
         self._drop_last = loader.drop_last
         self._index_sampler = loader._index_sampler
         self._num_workers = loader.num_workers
+        ws, rank = _get_distributed_settings()
```
Change 1: Move distributed information to the constructor of `DataLoaderIterator`.

ejguan commented on torch/utils/data/dataloader.py:
```diff
@@ -565,63 +556,16 @@ def _create_warning_msg(num_worker_suggest, num_worker_created, cpuset_checked):
+        self._pg = None
         if isinstance(self._dataset, IterDataPipe):
+            if dist.is_available() and dist.is_initialized():
+                self._pg = dist.new_group(backend="gloo")
```
Change 2: Use process group rather than store

ejguan replied:
It doesn't as `gloo` is required whenever PyTorch distributed is compiled.

NivekT commented on torch/utils/data/dataloader.py:
```diff
@@ -119,6 +117,13 @@ def _sharding_worker_init_fn(worker_init_fn, world_size, rank_id, worker_id):
         worker_init_fn(worker_id)
 
 
+def _share_dist_seed(generator, pg):
```
Question: so this handles everything that `_get_shared_seed` did? Including `seed_counter`, `timeout`, `check_interval`?

ejguan replied:
Yep. ProcessGroup itself contains `timeout` and `check_interval`. And, the counter would rely on `broadcast` function.

NivekT reviewed (approved):
LGTM! Just some questions:

I forgot where the existing tests are for these, can you point them out?

ejguan commented:
This PR has been validated by the newly introduced tests in https://github.com/pytorch/data/pull/727

VitalyFedyunin reviewed (approved):
By any chance, we can include a test case for it? (I especially concerned about testing same thing for DLV2)

ejguan commented:
> By any chance, we can include a test case for it? (I especially concerned about testing same thing for DLV2)

I was thinking about it. But, I add the tests for both DLv2 and DL in https://github.com/pytorch/data/pull/727 because I find the most of code are duplicated. WDYT? Should I move it to PyTorch Core?

VitalyFedyunin commented:
No need. Missed the fact that https://github.com/pytorch/data/pull/727 is also testing lazy init case.

ejguan committed:
[DataLoader] Make distributed lazily initialized & share seed via PG (#85279)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/84492

## Changes
- Move the logic of distributed sharding from the constructor of DataLoader to the constructor of DataLoaderIterator. This would prevent the Error caused by lazy distributed process initialization
- Replace distributed store by process group (`gloo`) to share the random seed because `mpi` backend doesn't provide distributed store.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85279

Differential Revision: D39625646

Pulled By: ejguan

fbshipit-source-id: 96436f0630ca53fda8720845084fb1d7a19c121f

ejguan commented:
@pytorchbot merge -g
